contracts/src/rewards/RewardsManager.sol

```diff
@@ -0,0 +1,122 @@
+import "../registry/NodeRegistry.sol";
+import "../epoch/EpochManager.sol";
+import "./RewardsAggregator.sol";
+import "../utils/MathUtils.sol";
+import {SD59x18, sd, intoInt256, intoUint256, UNIT, convert} from "prb/math/SD59x18.sol";
+
+/**
+ * @title Fleek Reward Contract
+ * @dev This contract calculates and distributes the rewards
+ */
+
+contract FleekReward is Controlled {
+    FleekToken private fleekToken;
+    RewardsAggregator public rewardsAggregator;
+    EpochManager public epochManager;
+    NodeRegistry public nodeRegistry;
+    SD59x18 public inflationInLastEpoch;
+
+    bool private initialized;
+
+    // epoch number => bool
+    mapping(uint256 => bool) public rewardsDistribution;
+
+    uint256 constant DAYS_IN_YEAR = 365;
+
+    // min factor by which the inflation can go down based on usage in %
+    SD59x18 public minInflationFactor;
+    // max inflation for the year in %
+    SD59x18 public maxInflation;
+    // price per byte
+    SD59x18 public price;
+    // Cost of running node
+    SD59x18 public cost;
+
+    event RewardMinted(address indexed account, uint256 amount);
+
+    function initialize(address _controller, address _token, address _epoch, address _aggregator, address _registry)
+        external
+    {
+        require(!initialized, "Rewards contract already initialized");
+        Controlled._init(_controller);
+        fleekToken = FleekToken(_token);
+        nodeRegistry = NodeRegistry(_registry);
+        epochManager = EpochManager(_epoch);
+        rewardsAggregator = RewardsAggregator(_aggregator);
+        inflationInLastEpoch = maxInflation;
+        initialized = true;
+    }
+
+    function setInflationRate(SD59x18 _inflationRate) external {
+        maxInflation = _inflationRate;
+    }
+
+    function setMinInflationFactor(SD59x18 _minInflationFactor) external {
+        minInflationFactor = _minInflationFactor;
+    }
+
+    function setPrice(SD59x18 _price) external {
+        price = _price;
+    }
+
+    function setCost(SD59x18 _cost) external {
+        cost = _cost;
+    }
+
+    /**
+     * @dev Distribute reward tokens to addresses.
+     * @param epoch epoch for which the rewards to be distributed
+     */
+    function distributeRewards(uint256 epoch) public onlyController {
+        require(epochManager.epoch() != epoch, "cannot distribute rewards for current epoch");
+        require(!rewardsDistribution[epoch], "rewards already distributed for this epoch");
+
+        SD59x18 _uActual = convert(int256(rewardsAggregator.getDataForEpoch(epoch)));
+        SD59x18 _uPotential = convert(int256(rewardsAggregator.getAvgUsageNEpochs(epoch)));
+
+        SD59x18 _totalMint = _getMintRate(_uActual, _uPotential);
+        // todo: variable distribution
+        // 75% goes to edge node
+        SD59x18 _toEdgeNode = _totalMint.mul(sd(0.75e18));
+        string[] memory publicKeys = rewardsAggregator.getPublicKeys(epoch);
+
+        for (uint256 i = 0; i < publicKeys.length; i++) {
```

Comment:
minor nit: Code be optimized a bit to decrease opcodes. Im fine merging in and optimizing later. Some things here in this for loop
- caching publicKeys.length to variable and check against that
- publicKeys.length array length could never be greater then max uint256 so we can do unchecked increment safely on i

Reply:
I have added these fixes. thanks for pointing it out.